packages/ckeditor5-ui/src/menubar/menubarmenulistview.ts

```diff
@@ -27,5 +32,64 @@ export default class MenuBarMenuListView extends ListView {
 		super( locale );
 
 		this.role = 'menu';
+		this.items.on( 'change', this._preallocateCheckSpacerInItemsIfNeeded.bind( this ) );
+	}
+
+	/**
+	 * This method adds empty space if there is any toggleable item in the list.
+	 * It makes the list properly aligned.
+	 */
+	private _preallocateCheckSpacerInItemsIfNeeded() {
```

Comment:
I understand we have some code duplication, because similar code already exists but it's in "dropdown" space, not "menubar" space.

This code here for some reason looks overcomplicated. I somehow like the other solution better and I wonder what's the difference here, as the structures are more-or-less similiar. In the "dropdowns code" we also iterate through the items list, but we only need to check this:

```javascript
item instanceof ListItemView && item.children.first instanceof ListItemButtonView
```

to decide if that's a button or not. Why it is much more complicated here (you need two helper functions)? I also like in the other function that you first compile the list of buttons, and then check if any is toggleable, and then check the property for all buttons. Less places where you use "weird" functions like `pickListButtonMenuViewIfPresent()`.

I also wonder if we have a performance problem here due to this being called on every `change` callback. Are all items added together, or are they added one by one?

Reply:
@scofalik There are a few things why it's so complicated and why I decided to not unify this behavior. At this moment, we have to:

1. Iterate over all items of menu and try to pick all list items buttons. If it's:
1.1. `ListItemButtonView` - then pick it as it is, it's safe to use `instanceof` here, because `ListItemButtonView` does not use `MenuBarMenuListView`
1.1. `ListSeparatorView`  - do nothing, rejected in `reduce`.
1.1. `MenuBarMenuView` there is circular reference, and we have to define a `isNestedMenuLikeView` guard that determines if it's something that has similar structure to `MenuBarMenuView` and has `buttonView`
2. If some of the buttons are checkboxes then force allocate list items space in all buttons.

As you see, it raises two problems:

1. Circular references in this file. We cannot use `instanceof MenuBarMenuView` because `MenuBarMenuView` imports this file itself. 
2. We cannot assume that the first item of `MenuBarMenuListItemView` is always the button. This view contains only `children` which can hold anything, so it's why there is filtering made. 

I think that this method should not be unified because menu is much more complex. The pending dropdown implementation implements it slightly differently too.

Reply:
I wanted to analyze this and propose something but honestly, there's so much complication (mess) in the menu bar that I gave up. I think this feature far too many classes which extends too broad classes that allow for "any children" and the extended classes do not narrow down the children list.

Reply:
@scofalik That's why dropdown menus operate mainly using trees